### ISSUE TITLE
Postico DNS not using PC DNS for private domains

### ISSUE THREAD

**Issue by tony-caffe** [jakob/Postico#609]
## What did you do?

- I clicked connect.
- Then I waited for Postico to look for my AWS RedShift via URL that is private only.
- It fails the first time since it doesnt use my local PC's DNS servers, seems to use its own provided ones.
- After I get the following error message, I click try again and it works just fine.

```
could not connect to server: Operation timed out
    Is the server running on host "myawsaddress.redshift.amazonaws.com" (34.xxx.xx.xx) and accepting
    TCP/IP connections on port 5439?
```

## What did you expect to happen?

- I expected Postico to use my local DNS that is pushed from my VPN

## What actually happened?

- Connection Timeout

## What software versions are you using?

Postico version:  
1.4.3

macOS version:  
10.14.1

PostgreSQL version:  
PostgreSQL 8.0.2 Redshift 1.0.4515

This happens for all employees at my company that use this app for RedShift. I think Postico should have the first try to connect to Local PC DNS Servers instead of the unknown ones Postco is actually using. I dont like that we dont know the DNS Postico uses on the first try, as we are wanting to keep our DNS requests secure and thru the VPN.

**Comment by tony-caffe** [comment]
Hi! Thank you for the response.

Yes Command line is what I normally use and I never have this issue and its as instant to connect as it is on the second try for Postico.

OpenVPN client on each user is pushed an internal DNS from Amazon and if for some reason that DNS cant find a resource it has DNS backup as google 8.8.8.8. This push only overwrites the end-users DNS on their Mac when connected and returns to the default, which usually points to their home router.

The terminal always outputs the correct DNS even with nslookup on the VPN. It is an odd behavior indeed.

Sorry for not knowing the workings of Libpg but does it call a specific DNS server and then has a fallback? Perhaps command line doesnt interface with libpg the same way.

**Comment by jakob** [comment]
There is no hard coded DNS server in Postico and there also isn't a hard coded DNS server in libpq.

psql uses libpq just like Postico, so both should end up with the same IP addresses for a host.

They just ask the OS to look up the name.

The only difference is that Postico effectively looks up the name twice (once just to make sure that the name exists, and then libpq looks up the name again). Maybe this causes issues with the VPN?

Let me try to recap the situation, just to make sure I understand correctly:

1) Your user connects to a VPN
2) The VPN provides a DNS server which returns different IPs for some public names
3) When you connect with Postico for the first time, it tries to connect to an IP address that must have been returned from a public DNS server
4) But subsequently it works again?

I'm a bit stumped and I don't have any ideas how to debug this at the moment...

Also, I don't really understand how the VPN part is supposed to work. Can you explain a bit more about your VPN? Where is the VPN? Why do the Redshift servers have a different IP when you connect through the VPN?

**Comment by tony-caffe** [comment]
1. Your user connects to a VPN
2. The VPN provides a DNS server which returns different IPs for some Private names on Private network (AWS Network to be exact)
3. When you connect with Postico for the first time, it tries to connect to an IP address (external IP lookup for an internal only IP) that must have been returned from a public DNS server
4. But subsequently it works again and finds the internal IP for this private resource

Mind you the AWS URL to RedShift is public and private but it has no access allowed publicly. With the VPN, it allows us to call the private IP (Dynamic IP). Thus why we need the VPN to push internal DNS server to allow internal only URLs and IPs to be resolved and accessible remotely.

**Comment by jakob** [comment]
OK, I think I understand now. You have a VPN server running inside the AWS network, and this allows you to connect to the internal IP of the Redshift cluster. This requires overriding the DNS server. But for some reason, this sometimes doesn't work with Postico, because you get the public IP instead of the internal one.

Do you run OpenVPN on an EC2 instance?

Could you share your OpenVPN configuration with us, so we can try to reproduce the issue?

Also, is there a reason why you don't use SSH tunnels instead of a VPN? SSH Tunnels work very reliably in Postico. With SSH, the database host name is always resolved on the SSH server side, so you won't have the DNS issue.

**Comment by tony-caffe** [comment]
For Scaleability, reliability and usability, a VPN is best. This doesnt appear to be a VPN issue though. Maybe its a pglib issue I should bring up to them. Seems odd. I dont have this issue with tools like MySQL workbench or URLs for internal Load Balancers in the browser, all which need to use the local DNS (pushed by VPN) so it can ping the AWS DNS server.

Also my VPN is EC2 but the server isnt hosting or providing the DNS. Amazon for our private network segments all require 1 DNS provided by AWS to interconnect, so we just take advantage of their already in place DNS and hardware resources by providing (pushing to) all the VPN clients its address to use to resolve DNS. EC2 server is literally a tunnel and means of authentication for our many users on many different OS and environments.

**Comment by jakob** [comment]
I made a little debug app that simulates how psql and Postico resolve host names. For all my hosts, this always returns the same IPs, but I assume that on your VPN, this would return different IPs.

Can you try it and report back with output for your hostnames? (You can redact parts that are private)

[DNSDebug.zip](https://github.com/jakob/Postico/files/2567833/DNSDebug.zip)

**Comment by tony-caffe** [comment]
Here is off of the VPN. It works as it should for both, looking up the public names since its not on the VPN's DNS:

```
Calling getaddrinfo(myAWSAddress.redshift.amazonaws.com)
(0) 34.206.xx.xx
1 result(s)
Calling CFHostCreateWithName(myAWSAddress.redshift.amazonaws.com)
(0) 34.206.xx.xx
1 result(s)
Calling getaddrinfo(myAWSAddress.redshift.amazonaws.com)
(0) 34.206.xx.xx
1 result(s)
```

On the VPN with my DNS changed by the VPN, the same results are above exactly.  Even on multiple tries, the External IP is only used.


Now if I use the command line on the VPN, here are the correct results. Notice the IP is an internal IP and so is the DNS server used to get the correct IP:

```
$ nslookup myAWSAddress.redshift.amazonaws.com
Server:		172.31.x.x
Address:	172.31.x.x#53

Non-authoritative answer:
myAWSAddress.redshift.amazonaws.com	canonical name = ec2-34-206-xx-xx.compute-1.amazonaws.com.
Name:	ec2-34-206-xx-xx.compute-1.amazonaws.com
Address: 172.31.xx.xxx
```

I commented out the information to avoid publicly placing my addresses and IPs.

**Comment by tony-caffe** [comment]
Cool tool BTW! 👍

**Comment by jakob** [comment]
Thanks for testing.

I guess what is happening is that macOS is caching DNS results. So if you've looked up `myAWSAddress.redshift.amazonaws.com` before connecting to the VPN, macOS somewhere caches the result. Then you connect to the VPN, your DNS server is changed, but calls to `getaddrinfo()` still return the cached result. Either the failed connection attempt, or the repeated name lookups must cause macOS to invalidate the cache, so that the second connection attempt succeeds.

The man page of `nslookup` contains the following notice:

> The nslookup command does not use the host name and address resolution or the DNS
       query routing mechanisms used by other processes running on macOS.  The results of
       name or address queries printed by nslookup may differ from those found by other
       processes that use the macOS native name and address resolution mechanisms.  The
       results of DNS queries may also differ from queries that use the macOS DNS routing
       library.

So `nslookup` is probably no good for debugging this issue.

The only thing that I can't explain is why connecting from the command line (using `psql`) does not suffer from this problem.

In any case, this seems to be an issue with OpenVPN / macOS. As far as I understand, `getaddrinfo()` should not return results from the old DNS server after a VPN connection has been established. OpenVPN should invalidate the local DNS cache, but it seems that it doesn't do that. Also, macOS should probably invalidate the cache when name server configuration is changed.

I'm not sure if there is anything I can do in Postico to fix this issue. How would Postico know that the system is returning an outdated IP for a hostname? And how would Postico tell the system to invalidate the cache?

**Comment by tony-caffe** [comment]
@jakob Thank you sir for your help. It makes sense that its a MacOS issue, since the OS will want to make connections seem faster by caching DNS. It is weird that other apps do not use MacOS cache results for things like MySQL Workbench etc.

I guess your app could force a DNS flush or to return a non-cached result, but that isnt really your apps concern nor is it the concern of psql. A change like this would only affect a few people that use this AWS environment like I do. Most dont care that their Redshift or other DBs are public facing, its the lack of management experience that allows external connections to all the various resources in AWS.

**Comment by jakob** [comment]
OK, so I tried setting up a Redshift Cluster.

If you select "Publicly Accessible: No" the cluster is not assigned a public IP address, and looking up the host name always resolves in the private IP address, even when using the public DNS server.

This would fix all your macOS/OpenVPN issues.

Why do your clusters even have a public IP if you do not want to allow external connections?

**Comment by tony-caffe** [comment]
We have a few outside contractors that use this and others that are outside of the VPN. Redshift is also linked with BI Tools that require access. Sadly it has to stay as both. Thats why the VPN DNS push is so useful and secure.

**Comment by jakob** [comment]
I'm closing this issue since as far as I can see it's working fine from our end.

**Comment by trunsky** [comment]
@tony-caffe did you ever find a solution to having the db publicly accessible and postico working with OpenVPN?

**Comment by tony-caffe** [comment]
@trunsky Yes I found a work around via split brain DNS and network route in AWS but psql client and MacOS dont look for the computer's DNS and routes initially. So typically DNS is cached on MacOS on restart of OS. It has to have a failed attempt first to then clear the cache and use OpenVPN DNS and then will use the internal route. I opened an issue with MacOS but this is a feature they use to make MacOS feel faster (Hence the use of cache).
